### PR TITLE
[!!!] Implement TYPO3 v11 & v12 support and drop support for TYPO3 v6, v7, v8 & v9

### PR DESCRIPTION
**Pull request by djcprinse**
### This release will be breaking and bring the extension to v2.0.0.

The required work for TYPO3 v11 and v12 support has been initially done by running rector against the extension. This resulted in the new Icons API file being added.

Since rector could not implement the PSR-14 events for FAL files, which were introduced in TYPO3 v10.2, these have been implemented manually. The `RemoteObjectUpdate` signal slot class has been refactored and was moved from the `CacheControl` namespace to `Resource\Event`. The class now implements event listener methods

The `ClearCacheActionsHookInterface`, which was implemented in the ToolbarItem class, has been deprecated in TYPO3 v11 and had to be refactored to a PSR-14 event too. The `FlushCacheActionEvent` class implements the new `ModifyClearCacheActionsEvent` event listener and adds the cache action button to the action list. In TYPO3 v10, the old `ToolbarItem` class implementation is being used. The `ToolbarItem` class now extends the new action class and shares the same code for configuring the button.

The `ImagePreviewConfiguration` signal slot class was incorrectly being used in TYPO3 v9 and up. This class was added for TYPO3 v8 and below, because in that version preview images were incorrectly being generated without any dimensions. This resulted in multiple and different preview images being generated. This signal slot has been removed since it is no longer being used.

Because we implemented the `ImageDimensionsExtraction` extractor in v1.12.0, which makes sure the image dimensions are being extracted after uploading an image, the RecordMonitor hook which was necessary in earlier TYPO3 versions to correctly fetch image dimensions, could be removed. TYPO3 v8+ uses the `ImageDimensionsExtraction` extractor correctly and updates the image metadata width and height on upload of a file.

Where possible, strict typing has been added to classes, methods and properties.

### Custom endpoint option
The option which would be added in pull request #63 has been merged to this feature branch, since both pull request would require a new major version.

### CODE REVIEW

File: Classes/Driver/AmazonS3Driver.php [hunk 1]
```diff
@@ -778,7 +817,7 @@ public function copyFolderWithinStorage($sourceFolderIdentifier, $targetFolderId
      * @return string The file contents
      * @throws InvalidPathException
      */
-    public function getFileContents($fileIdentifier)
+    public function getFileContents($fileIdentifier): string
```
Comment (DerFrenk) [inline Classes/Driver/AmazonS3Driver.php:820]: De gereturnde `file_get_contents` geeft `false` terug indien iets niet opgehaald kan worden. Zelfde geldt voor `file_put_contents` hieronder

File: Classes/Driver/AmazonS3Driver.php [hunk 2]
```diff
@@ -1139,7 +1180,7 @@ public function getFoldersInFolder(
         array $folderNameFilterCallbacks = [],
         $sort = '',
         $sortRev = false
-    ) {
+    ): array {
         if ($start === false && $numberOfItems === false) {
```
Comment (DerFrenk) [inline Classes/Driver/AmazonS3Driver.php:1184]: In hoeverre moeten we (nu al) iets met strict typing qua functie params? We definieren `$start = 0` en controleren vervolgens of ie niet `false` is. Beetje lelijk dit zo

File: Classes/Driver/AmazonS3Driver.php [hunk 3]
```diff
@@ -622,7 +642,7 @@ public function deleteFile($fileIdentifier)
      * @return string
      * @throws InvalidPathException
      */
-    public function hash($fileIdentifier, $hashAlgorithm)
+    public function hash($fileIdentifier, $hashAlgorithm): string
```
Comment (DerFrenk) [inline Classes/Driver/AmazonS3Driver.php:645]: `sha1_file` en `md5_file` kunnen false teruggeven

File: Classes/Driver/AmazonS3Driver.php (hunk 1 above)
Comment (djcprinse) [reply Classes/Driver/AmazonS3Driver.php:834]: Aangepast door `return file_get_contents() ?: ''` te gebruiken in `getFileContents`.

In `setFileContents` is er een exception toegevoegd op het moment dat `file_put_contents` false als waarde teruggeeft.

File: Classes/Driver/AmazonS3Driver.php (hunk 2 above)
Comment (djcprinse) [reply Classes/Driver/AmazonS3Driver.php:1204]: Dit gaat niet vanwege de `AbstractDriverInterface` waar return types ontbreken

File: Classes/Driver/AmazonS3Driver.php (hunk 3 above)
Comment (djcprinse) [reply Classes/Driver/AmazonS3Driver.php:647]: Dit is verholpen door een exception op te gooien als `sha1_file` of `md5_file` false als waarde heeft.

File: ext_localconf.php [hunk 4]
```diff
@@ -140,7 +78,6 @@
         // phpcs:enable
     }
 
-
     if (class_exists(\TYPO3\CMS\Core\Information\Typo3Version::class)) {
```
Comment (RinyVT) [inline ext_localconf.php:81]: This check is not needed anymore since the `Typo3Version` class is always available now.

**Review by DerFrenk — APPROVED** [review]

**Pull request merged**